File: backend/ECommerce.Application/Services/DashboardService.cs
namespace ECommerce.Application;

public class DashboardService : IDashboardService
{
    private readonly IDashboardRepository _dashboardRepository;

    public DashboardService(IDashboardRepository dashboardRepository)
    {
        _dashboardRepository = dashboardRepository;
    }

    public async Task<DashboardSummaryResponse> GetSummaryAsync(DateTime? startDate, DateTime? endDate)
    {
        var start = startDate ?? DateTime.UtcNow.AddDays(-30);
        var end = endDate ?? DateTime.UtcNow;

        var data = await _dashboardRepository.GetRawDataAsync(start, end);

        var revenueGrowth = data.PreviousPeriodRevenue > 0
            ? (int)((data.CurrentPeriodRevenue - data.PreviousPeriodRevenue) / data.PreviousPeriodRevenue * 100)
            : 0;

        var orderGrowth = data.PreviousPeriodOrders > 0
            ? (int)((data.CurrentPeriodOrders - data.PreviousPeriodOrders) / (decimal)data.PreviousPeriodOrders * 100)
            : 0;

        return new DashboardSummaryResponse(
            data.CurrentPeriodRevenue,
            data.CurrentPeriodOrders,
            data.TotalCustomers,
            data.TotalActiveProducts,
            data.PendingOrders,
            revenueGrowth,
            orderGrowth
        );
    }

    public async Task<RevenueReportResponse> GetRevenueReportAsync(DateTime? startDate, DateTime? endDate)
    {
        var start = startDate ?? DateTime.UtcNow.AddDays(-30);
        var end = endDate ?? DateTime.UtcNow;

        var data = await _dashboardRepository.GetRawDataAsync(start, end);

        // Daily revenue
        var dailyRevenue = data.OrdersWithItems
            .GroupBy(o => o.CreatedAt.Date)
            .Select(g => new RevenueByDate(
                g.Key,
                g.Sum(o => o.TotalAmount),
                g.Count()
            ))
            .OrderBy(r => r.Date)
            .ToList();

        // Revenue by category
        var categoryRevenue = data.OrdersWithItems
            .SelectMany(o => o.Items)
            .GroupBy(i => i.CategoryName)
            .Select(g => new RevenueByCategory(
                g.Key,
                g.Sum(i => i.UnitPrice * i.Quantity),
                g.Count(),
                0
            ))
            .OrderByDescending(c => c.Revenue)
            .ToList();

        var totalCategoryRevenue = categoryRevenue.Sum(c => c.Revenue);
        categoryRevenue = categoryRevenue.Select(c => c with
        {
            Percentage = totalCategoryRevenue > 0 ? (double)(c.Revenue / totalCategoryRevenue * 100) : 0
        }).ToList();

        var totalRevenue = data.CurrentPeriodRevenue;
        var averageOrderValue = data.CurrentPeriodOrders > 0 ? totalRevenue / data.CurrentPeriodOrders : 0;

        return new RevenueReportResponse(
            dailyRevenue,
            categoryRevenue,
            totalRevenue,
            averageOrderValue
        );
    }

    public async Task<TopProductsResponse> GetTopProductsAsync(int limit, DateTime? startDate, DateTime? endDate)
    {
        var start = startDate ?? DateTime.UtcNow.AddDays(-30);
        var end = endDate ?? DateTime.UtcNow;

        var data = await _dashboardRepository.GetRawDataAsync(start, end);

        var topProducts = data.OrdersWithItems
            .SelectMany(o => o.Items)
            .GroupBy(i => new { i.ProductId, i.ProductName })
            .Select(g => new TopProduct(
                g.Key.ProductId,
                g.Key.ProductName,
                g.Sum(i => i.Quantity),
                g.Sum(i => i.UnitPrice * i.Quantity),
                g.First().ImageUrl
            ))
            .OrderByDescending(p => p.TotalSold)
            .Take(limit)
            .ToList();

        return new TopProductsResponse(topProducts);
    }

    public async Task<InventoryReportResponse> GetInventoryReportAsync()
    {
        var stockData = await _dashboardRepository.GetProductStockDataAsync();

        var totalVariants = stockData.Count;
        var inStock = stockData.Count(v => v.StockQuantity > 10);
        var lowStock = stockData.Count(v => v.StockQuantity > 0 && v.StockQuantity <= 10);
        var outOfStock = stockData.Count(v => v.StockQuantity == 0);

        var lowStockProducts = stockData
            .Where(v => v.StockQuantity <= 10)
            .Select(v => new LowStockProduct(
                v.VariantId,
                v.ProductName,
                v.VariantName,
                v.StockQuantity,
                v.Sku
            ))
            .ToList();

        return new InventoryReportResponse(
            totalVariants,
            inStock,
            lowStock,
            outOfStock,
            lowStockProducts
        );
    }
}

File: frontend/src/app/features/admin/dashboard/dashboard.service.ts
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface DashboardSummary {
  totalRevenue: number;
  totalOrders: number;
  totalCustomers: number;
  totalProducts: number;
  pendingOrders: number;
  revenueGrowthPercent: number;
  orderGrowthPercent: number;
}

export interface RevenueByDate {
  date: string;
  revenue: number;
  orderCount: number;
}

export interface RevenueByCategory {
  categoryName: string;
  revenue: number;
  orderCount: number;
  percentage: number;
}

export interface RevenueReport {
  dailyRevenue: RevenueByDate[];
  categoryRevenue: RevenueByCategory[];
  totalRevenue: number;
  averageOrderValue: number;
}

export interface TopProduct {
  productId: number;
  productName: string;
  totalSold: number;
  revenue: number;
  imageUrl: string | null;
}

export interface TopProductsResponse {
  products: TopProduct[];
}

export interface LowStockProduct {
  variantId: number;
  productName: string;
  variantName: string;
  stockQuantity: number;
  sku: string;
}

export interface InventoryReport {
  totalVariants: number;
  inStock: number;
  lowStock: number;
  outOfStock: number;
  lowStockProducts: LowStockProduct[];
}

@Injectable({ providedIn: 'root' })
export class DashboardService {
  private readonly apiUrl = 'http://localhost:5159/api/v1/dashboard';

  constructor(private readonly http: HttpClient) {}

  private headers(token: string): HttpHeaders {
    return new HttpHeaders({ Authorization: `Bearer ${token}` });
  }

  private dateParams(startDate?: string, endDate?: string): HttpParams {
    let params = new HttpParams();
    if (startDate) params = params.set('startDate', startDate);
    if (endDate) params = params.set('endDate', endDate);
    return params;
  }

  getSummary(token: string, startDate?: string, endDate?: string): Observable<DashboardSummary> {
    return this.http.get<DashboardSummary>(`${this.apiUrl}/summary`, {
      headers: this.headers(token),
      params: this.dateParams(startDate, endDate),
    });
  }

  getRevenueReport(token: string, startDate?: string, endDate?: string): Observable<RevenueReport> {
    return this.http.get<RevenueReport>(`${this.apiUrl}/revenue`, {
      headers: this.headers(token),
      params: this.dateParams(startDate, endDate),
    });
  }

  getTopProducts(
    token: string,
    limit = 10,
    startDate?: string,
    endDate?: string,
  ): Observable<TopProductsResponse> {
    const params = this.dateParams(startDate, endDate).set('limit', limit);
    return this.http.get<TopProductsResponse>(`${this.apiUrl}/top-products`, {
      headers: this.headers(token),
      params,
    });
  }

  getInventoryReport(token: string): Observable<InventoryReport> {
    return this.http.get<InventoryReport>(`${this.apiUrl}/inventory`, {
      headers: this.headers(token),
    });
  }
}